core/arch/x86/proc.c

```diff
@@ -358,39 +358,40 @@ proc_init_arch(void)
-         * Fig 13-1: 1) cpuid OSXSAVE; 2) xgetbv 0x6; 3) cpuid AVX.
-         * Xref i#1278, i#1030, i#437.
-         */
+    if (proc_has_feature(FEATURE_OSXSAVE)) {
+        uint bv_high = 0, bv_low = 0;
         dr_xgetbv(&bv_high, &bv_low);
-        LOG(GLOBAL, LOG_TOP, 2, "\txgetbv => 0x%08x%08x\n", bv_high, bv_low);
-        if (TESTALL(XCR0_AVX | XCR0_SSE, bv_low)) {
-            avx_enabled = true;
-            LOG(GLOBAL, LOG_TOP, 1, "\tProcessor and OS fully support AVX\n");
-        } else {
-            LOG(GLOBAL, LOG_TOP, 1, "\tOS does NOT support AVX\n");
-        }
-    }
-    if (proc_has_feature(FEATURE_AVX512) && proc_has_feature(FEATURE_OSXSAVE)) {
-        CLIENT_ASSERT(proc_has_feature(FEATURE_AVX), "Processor has AVX-512 but no AVX?");
-        if (TESTALL(XCR0_HI16_ZMM | XCR0_ZMM_HI256 | XCR0_OPMASK, bv_low)) {
-            /* XXX i#1312: It is unclear whether the kernel uses CR0 bits to disable
-             * AVX-512 for its own lazy context switching optimization. Experimental
-             * testing hasn't observed this so far. But if it does, then our lazy context
-             * switch would interfere with the kernel's and more support needs to be
-             * added. A possible solution could be to run small dedicated AVX-512
-             * instructions in decode itself if the processor and OS supports it. But how
-             * would we know if support by the OS is supposed to be there?
+        if (proc_has_feature(FEATURE_AVX)) {
+            /* Even if the processor supports AVX, it will #UD on any AVX instruction
+             * if the OS hasn't enabled YMM and XMM state saving.
+             * To check that, we invoke xgetbv -- for which we need FEATURE_OSXSAVE.
+             * FEATURE_OSXSAVE is also listed as one of the 3 steps in Intel Vol 1
+             * Fig 13-1: 1) cpuid OSXSAVE; 2) xgetbv 0x6; 3) cpuid AVX.
+             * Xref i#1278, i#1030, i#437.
              */
-            avx512_enabled = true;
-            LOG(GLOBAL, LOG_TOP, 1, "\tProcessor and OS fully support AVX-512\n");
-        } else {
-            LOG(GLOBAL, LOG_TOP, 1, "\tOS does NOT support AVX-512\n");
+            LOG(GLOBAL, LOG_TOP, 2, "\txgetbv => 0x%08x%08x\n", bv_high, bv_low);
+            if (TESTALL(XCR0_AVX | XCR0_SSE, bv_low)) {
+                avx_enabled = true;
+                LOG(GLOBAL, LOG_TOP, 1, "\tProcessor and OS fully support AVX\n");
+            } else {
+                LOG(GLOBAL, LOG_TOP, 1, "\tOS does NOT support AVX\n");
+            }
+        }
+        if (proc_has_feature(FEATURE_AVX512)) {
+            if (TESTALL(XCR0_HI16_ZMM | XCR0_ZMM_HI256 | XCR0_OPMASK, bv_low)) {
+                /* XXX i#1312: It is unclear whether the kernel uses CR0 bits to disable
+                 * AVX-512 for its own lazy context switching optimization. Experimental
+                 * testing hasn't observed this so far. But if it does, then our lazy
```

Comment:
We have an answer to this now and can update this comment, right?

Reply:
Good idea. Done.